server/util/terminal/terminal.go

```diff
@@ -1,26 +1,42 @@
 package terminal
 
 import (
+	"strings"
+
 	bkterminal "github.com/buildkite/terminal-to-html/v3"
 )
 
 type ScreenWriter struct {
-	screen *bkterminal.Screen
-}
-
-func NewScreenWriter() *ScreenWriter {
-	return &ScreenWriter{screen: bkterminal.NewScreen()}
+	*bkterminal.Screen
+	ScrollOut         strings.Builder
+	ScrollOutWriteErr error
 }
 
-func (w *ScreenWriter) Write(p []byte) (int, error) {
-	bkterminal.ParseANSIToScreen(w.screen, p)
-	return len(p), nil
+// NewScreenWriter returns a ScreenWriter backed by an ANSI state machine with a
+// window size of windowHeight. Any lines that scroll off the top are written to
+// the ScrollOut string builder, and any write errors encountered during that
+// process are recorded in ScrollOutWriteErr.
+// A windowHeight of less than 1 indicates a window of unlimited size.
+func NewScreenWriter(windowHeight int) (*ScreenWriter, error) {
+	s, err := bkterminal.NewScreen(bkterminal.WithMaxSize(0, windowHeight), bkterminal.WithANSIRenderer())
+	if err != nil {
+		return nil, err
+	}
+	w := &ScreenWriter{Screen: s}
+	if windowHeight > 0 {
+		s.ScrollOutFunc = func(line string) { _, w.ScrollOutWriteErr = w.ScrollOut.WriteString(line) }
+	}
+	return w, nil
 }
 
-func (w *ScreenWriter) Render() []byte {
-	return w.screen.AsANSI()
+func (w *ScreenWriter) Render() string {
+	s, _ := w.Screen.AsANSI()
+	return s
 }
 
-func (w *ScreenWriter) PopExtraLines(numLinesToRetain int) []byte {
-	return w.screen.FlushLinesFromTop(numLinesToRetain)
+func (w *ScreenWriter) Reset(windowHeight int) error {
+	w.ScrollOut.Reset()
```

Comment:
i don't know what i want to call this--outputaccumulator?  something in that vein? i just find it difficult to follow what's going on when this is accessed by code in `eventlog`--it makes it read like `ScrollOutFunc` isn't abstracted away for the caller even though it (mostly) is.

Reply:
I like OutputAccumulator; I changed it to that.